analyses/cms-open-data-ttbar/ml.py

```diff
@@ -0,0 +1,244 @@
+import ROOT
+import os
+from dataclasses import dataclass
+
+# histogram bin lower limit to use for each ML input feature
+bin_low = [0, 0, 0, 0, 50, 50, 50, 50, 25, 25, 25, 25, 0, 0, 0, 0, -1, -1, -1, -1]
+
+# histogram bin upper limit to use for each ML input feature
+bin_high = [6, 6, 6, 6, 300, 300, 550, 550, 300, 300, 300, 300, 1, 1, 1, 1, 1, 1, 1, 1]
+
+# names of each ML input feature (used when creating histograms)
+feature_names = [
+    "deltar_leptonbtoplep", "deltar_w1w2", "deltar_w1btophad", "deltar_w2btophad",
+    "mass_leptonbtoplep",   "mass_w1w2",   "mass_w1w2btophad", "pt_w1w2btophad",
+    "pt_w1",                "pt_w2",       "pt_btophad",       "pt_btoplep", 
+    "btag_w1",              "btag_w2",     "btag_btophad",     "btag_btoplep", 
+    "qgl_w1",               "qgl_w2",      "qgl_btophad",      "qgl_btoplep",
+]
+
+# labels for each ML input feature (used for plotting)
+feature_labels = [
+    "Delta R between b_{top-lep} Jet and Lepton",
+    "Delta R between the two W Jets",
+    "Delta R between first W Jet and b_{top-had} Jet",
+    "Delta R between second W Jet and b_{top-had} Jet",
+    "Combined Mass of b_{top-lep} Jet and Lepton [GeV]",
+    "Combined Mass of the two W Jets [GeV]",
+    "Combined Mass of b_{top-had} Jet and the two W Jets [GeV]",
+    "Combined p_T of b_{top-had} Jet and the two W Jets [GeV]",
+    "p_T of the first W Jet [GeV]",
+    "p_T of the second W Jet [GeV]",
+    "p_T of the b_{top-had} Jet [GeV]",
+    "p_T of the b_{top-lep} Jet [GeV]",
+    "btagCSVV2 of the first W Jet",
+    "btagCSVV2 of the second W Jet",
+    "btagCSVV2 of the b_{top-had} Jet",
+    "btagCSVV2 of the b_{top-lep} Jet",
+    "Quark vs Gluon likelihood discriminator of the first W Jet",
+    "Quark vs Gluon likelihood discriminator of the second W Jet",
+    "Quark vs Gluon likelihood discriminator of the b_{top-had} Jet",
+    "Quark vs Gluon likelihood discriminator of the b_{top-lep} Jet",
+]
+
+@dataclass
+class MLHistoConf:
+  name: str
+  title: str
+  binning: (float, float, int) # nbins, low, high
+
+ml_features_config: list[MLHistoConf] = [
+    MLHistoConf(name = feature_names[i], title = feature_labels[i], binning = (25, bin_low[i], bin_high[i])) for i in range(len(feature_names))
+]
+
+
+
+
+def compile_mlhelpers_cpp(fastforest_path, max_n_jets=6):
```

Comment:
This function (and in general all functions where `gInterpreter` or `gSystem` are called) need to be treated with special care in order to make them work on a distributed system. Calling into the interpreter or the modifying something on the local filesystem is not automatically propagated to all remote processes (and rightfully so). Functions that do this need to be wrapped in a way similar to what you see in the main script, e.g. at https://github.com/root-project/analysis-grand-challenge/blob/2715c0eeac4f192ec63893546775e99e9c563f5c/analyses/cms-open-data-ttbar/analysis.py#L330  ). You can already test this on a single machine with a Dask `LocalCluster` for example